homeassistant/components/tplink_omada/switch.py

```diff
@@ -0,0 +1,129 @@
+"""Support for TPLink Omada device toggle options."""
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from typing import Any
+
+from tplink_omada_client.definitions import PoEMode
+from tplink_omada_client.devices import OmadaSwitch, OmadaSwitchPortDetails
+from tplink_omada_client.omadaclient import OmadaClient, SwitchPortOverrides
+
+from homeassistant.components.switch import SwitchEntity, SwitchEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import OmadaCoordinator
+from .entity import OmadaSwitchDeviceEntity
+
+POE_SWITCH_ICON = "mdi:ethernet"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up switches."""
+    omada_client: OmadaClient = hass.data[DOMAIN][config_entry.entry_id]
+
+    # Naming fun. Omada switches, as in the network hardware
+    network_switches = await omada_client.get_switches()
+
+    entities: list = []
+    for switch in [
+        ns for ns in network_switches if ns.device_capabilities.supports_poe
+    ]:
+
+        def make_update_func(
+            network_switch: OmadaSwitch,
+        ) -> Callable[[OmadaClient], Awaitable[dict[str, OmadaSwitchPortDetails]]]:
+            async def update_func(
+                client: OmadaClient,
+            ) -> dict[str, OmadaSwitchPortDetails]:
+                ports = await client.get_switch_ports(network_switch)
+                return {p.port_id: p for p in ports}
+
+            return update_func
+
+        coordinator = OmadaCoordinator[OmadaSwitchPortDetails](
+            hass, omada_client, make_update_func(switch)
+        )
+
+        await coordinator.async_config_entry_first_refresh()
+
+        for idx, port_id in enumerate(coordinator.data):
+            if idx < switch.device_capabilities.poe_ports:
+                entities.append(
+                    OmadaNetworkSwitchPortPoEControl(
+                        coordinator, switch, omada_client, port_id
+                    )
+                )
+
+    async_add_entities(entities)
+
+
+def get_port_base_name(port: OmadaSwitchPortDetails) -> str:
+    """Get display name for a switch port."""
+
+    if port.name == f"Port{port.port}":
+        return f"Port {port.port}"
+    return f"Port {port.port} ({port.name})"
+
+
+class OmadaNetworkSwitchPortPoEControl(OmadaSwitchDeviceEntity, SwitchEntity):
+    """Representation of a PoE control toggle on a single network port on a switch."""
+
+    def __init__(
+        self,
+        coordinator: OmadaCoordinator[OmadaSwitchPortDetails],
+        device: OmadaSwitch,
+        omada_client: OmadaClient,
```

Comment:
No need to pass in the client, it's already available as `self.coordinator.omada_client`

Reply:
Removed